src/server/serverinventorymgr.cpp

```diff
@@ -0,0 +1,181 @@
+	} break;
+	default:
+		sanity_check(false); // abort
+		break;
+	}
+	return NULL;
+}
+
+void ServerInventoryMgr::setInventoryModified(const InventoryLocation &loc)
+{
+	switch (loc.type) {
+	case InventoryLocation::UNDEFINED:
+		break;
+	case InventoryLocation::PLAYER: {
+
+		RemotePlayer *player = m_env->getPlayer(loc.name.c_str());
+
+		if (!player)
+			return;
+
+		player->setModified(true);
+		player->inventory.setModified(true);
+		// Updates are sent in ServerEnvironment::step()
+	} break;
+	case InventoryLocation::NODEMETA: {
+		MapEditEvent event;
+		event.type = MEET_BLOCK_NODE_METADATA_CHANGED;
+		event.p = loc.p;
+		m_env->getMap().dispatchEvent(event);
+	} break;
+	case InventoryLocation::DETACHED: {
+		// Updates are sent in ServerEnvironment::step()
+	} break;
+	default:
+		sanity_check(false); // abort
+		break;
+	}
+}
+
+Inventory *ServerInventoryMgr::createDetachedInventory(
+		const std::string &name, IItemDefManager *idef, const std::string &player)
+{
+	if (m_detached_inventories.count(name) > 0) {
+		infostream << "Server clearing detached inventory \"" << name << "\""
+			   << std::endl;
+		delete m_detached_inventories[name];
+	} else {
+		infostream << "Server creating detached inventory \"" << name << "\""
+			   << std::endl;
+	}
+
+	Inventory *inv = new Inventory(idef);
+	sanity_check(inv);
+	m_detached_inventories[name] = inv;
+	if (!player.empty()) {
+		m_detached_inventories_player[name] = player;
+
+		// if player is connected, send him the inventory
+		if (RemotePlayer *p = m_env->getPlayer(name.c_str()))
+			m_env->getGameDef()->sendDetachedInventory(
+					inv, name, p->getPeerId());
+	} else {
+		// Inventory is for everybody, broadcast
+		m_env->getGameDef()->sendDetachedInventory(inv, name, PEER_ID_INEXISTENT);
+	}
+
+	return inv;
+}
+
+bool ServerInventoryMgr::removeDetachedInventory(const std::string &name)
+{
+	const auto &inv_it = m_detached_inventories.find(name);
+	if (inv_it == m_detached_inventories.end())
+		return false;
+
+	delete inv_it->second;
+	m_detached_inventories.erase(inv_it);
+
+	if (!m_env) // Mods are not done loading
+		return true;
+
+	const auto &player_it = m_detached_inventories_player.find(name);
+	if (player_it != m_detached_inventories_player.end()) {
+		RemotePlayer *player = m_env->getPlayer(player_it->second.c_str());
+
+		if (player && player->getPeerId() != PEER_ID_INEXISTENT)
+			m_env->getGameDef()->sendDetachedInventory(
+					nullptr, name, player->getPeerId());
+
+		m_detached_inventories_player.erase(player_it);
+	} else {
+		// Notify all players about the change
+		m_env->getGameDef()->sendDetachedInventory(
+				nullptr, name, PEER_ID_INEXISTENT);
+	}
+	return true;
+}
+
+void ServerInventoryMgr::applyOnInventories(std::function<void(ServerInventoryMgr *imgr,
+				const std::string &, Inventory *)>
+				apply_cb)
+{
+	for (const auto &detached_inventory : m_detached_inventories) {
+		apply_cb(this, detached_inventory.first, detached_inventory.second);
```

Comment:
I thought you wanted to include the `incremental` checks there, hence the suggestion for a callback.
Now it's no better than a simple map getter function like `const auto &inventories = m_inventory_mgr.getInventories();`.
If you would like to use callbacks anway (despite being kinda useless right now), it might not be needed to pass `this`. That parameter is already known by the caller (i.e. is superfluous).

Reply:
at the beginning i wanted to performing this sending from the manager, then i found a more generic way to apply things without taking the ownership of the map while searching for a good function name. I can restore the incremental behaviour if you prefer, yes